src/main/java/com/artipie/asto/LoggingStorage.java

```diff
@@ -0,0 +1,162 @@
+     *
+     * @param level Logging level.
+     * @param storage Delegate storage.
+     */
+    public LoggingStorage(final Level level, final Storage storage) {
+        this.level = level;
+        this.storage = storage;
+    }
+
+    @Override
+    public CompletableFuture<Boolean> exists(final Key key) {
+        return this.storage.exists(key).thenApply(
+            result -> {
+                this.log("Exists '%s': %s", key.string(), result);
+                return result;
+            }
+        );
+    }
+
+    @Override
+    public CompletableFuture<Collection<Key>> list(final Key prefix) {
+        return this.storage.list(prefix).thenApply(
+            result -> {
+                this.log("List '%s': %s", prefix.string(), result.size());
+                return result;
+            }
+        );
+    }
+
+    @Override
+    public CompletableFuture<Void> save(final Key key, final Content content) {
+        return this.storage.save(key, content).thenApply(
+            result -> {
+                this.log("Save '%s': %s", key.string(), content.size());
+                return result;
+            }
+        );
+    }
+
+    @Override
+    public CompletableFuture<Void> move(final Key source, final Key destination) {
+        return this.storage.move(source, destination).thenApply(
+            result -> {
+                this.log("Move '%s' '%s'", source.string(), destination.string());
+                return result;
+            }
+        );
+    }
+
+    @Override
+    public CompletableFuture<Long> size(final Key key) {
+        return this.storage.size(key).thenApply(
+            result -> {
+                this.log("Size '%s': %s", key.string(), result);
+                return result;
+            }
+        );
+    }
+
+    @Override
+    public CompletableFuture<Content> value(final Key key) {
+        return this.storage.value(key).thenApply(
+            result -> {
+                this.log("Value '%s': %s", key.string(), result.size());
+                return result;
+            }
+        );
+    }
+
+    @Override
+    public CompletableFuture<Void> delete(final Key key) {
+        return this.storage.delete(key).thenApply(
+            result -> {
+                this.log("Delete '%s'", key.string());
+                return result;
+            }
+        );
+    }
+
+    @Override
+    public CompletableFuture<Transaction> transaction(final List<Key> keys) {
+        return this.storage.transaction(keys).thenApply(
+            result -> {
+                this.log(
+                    "Transaction '%s'",
+                    keys.stream().map(Key::string).collect(Collectors.joining(", "))
+                );
+                return result;
+            }
+        );
+    }
+
+    /**
+     * Log message.
+     *
+     * @param msg The text message to be logged.
+     * @param args Optional arguments for string formatting.
+     */
+    private void log(final String msg, final Object... args) {
```

Comment:
@paulodamaso yes, it is also one line, but logic about logging level and target class is going to be duplicated. I have extracted a method for it to avoid logic duplication

Reply:
@olegmoz we won't be duplicating any logic, just accessing the defined fields.

Reply:
@paulodamaso bu the fact that we are accessing these fields IS logic. The fact that we are not using some other log level and using `this.storage` instead of `this` as log target - this is what captured in this method